### PR TITLE
[Coverage] Avoid emitting duplicate coverage mappings

### PR DESCRIPTION
**Pull request by vedantk**
- Only instrument closures and nested functions once.
- Only visit functions to build coverage mappings once (and assert that we don't create duplicate coverage mappings).

Addresses rdar://39200851 and rdar://39297172.

### CODE REVIEW

**Comment by vedantk** [comment]
@swift-ci Please smoke test

**Comment by vedantk** [comment]
@swift-ci Please Test Source Compatibility

**Comment by vedantk** [comment]
@swift-ci Please smoke test

**Comment by vedantk** [comment]
@swift-ci Please smoke test

**Comment by vedantk** [comment]
@swift-ci Please Test Source Compatibility

**Comment by vedantk** [comment]
@swift-ci Please Test Source Compatibility

File: lib/SIL/SILProfiler.cpp [hunk 1]
```diff
@@ -185,8 +200,11 @@ struct MapRegionCounters : public ASTWalker {
   std::pair<bool, Expr *> walkToExprPre(Expr *E) override {
     if (auto *IE = dyn_cast<IfExpr>(E)) {
       CounterMap[IE->getThenExpr()] = NextCounter++;
-    } else if (isa<AutoClosureExpr>(E) || isa<ClosureExpr>(E)) {
```
Comment (slavapestov) [inline lib/SIL/SILProfiler.cpp]: After type checking, AutoClosureExpr and ClosureExpr are the same semantically, so whatever you do with one should be done with the other.

File: lib/SILGen/SILGen.cpp [hunk 2]
```diff
@@ -483,6 +483,10 @@ static bool hasSILBody(FuncDecl *fd) {
   return fd->getBody(/*canSynthesize=*/false);
 }
 
+static bool haveProfiledAssociatedFuncDecl(SILDeclRef constant) {
+  return constant.isDefaultArgGenerator() || constant.isForeign;
```
Comment (slavapestov) [inline lib/SILGen/SILGen.cpp]: What about stored property initializers?

File: lib/SIL/SILCoverageMap.cpp [hunk 3]
```diff
@@ -40,6 +41,13 @@ SILCoverageMap::create(SILModule &M, StringRef Filename, StringRef Name,
...
   CM->Expressions = M.allocateCopy(Expressions);
 
+  // Assert that this coverage map is unique.
+  assert(llvm::none_of(M.coverageMaps,
+                       [&](const SILCoverageMap &OtherCM) {
+                         return OtherCM.PGOFuncName == CM->PGOFuncName;
+                       }) &&
+         "Duplicate coverage mapping for function");
+
   M.coverageMaps.push_back(CM);
```
Comment (slavapestov) [inline lib/SIL/SILCoverageMap.cpp]: Why not use a DenseMap instead? The assertion is O(n)...

File: lib/SIL/SILProfiler.cpp (hunk 1 above)
Comment (vedantk) [reply lib/SIL/SILProfiler.cpp]: I'll try getting rid of the distinction between autoclosures and closures here and see if anything breaks.

File: lib/SILGen/SILGen.cpp (hunk 2 above)
Comment (vedantk) [reply lib/SILGen/SILGen.cpp]: I'm not sure. Isn't the decl associated with a stored property initializer some vardecl? If so, we wouldn't be at risk of instrumenting it twice.

File: lib/SIL/SILCoverageMap.cpp (hunk 3 above)
Comment (vedantk) [reply lib/SIL/SILCoverageMap.cpp]: I thought the PR would be cleaner without this change. I'm planning on doing this as a follow-up, but can do it now if you'd prefer.

File: lib/SIL/SILProfiler.cpp (hunk 1 above)
Comment (vedantk) [reply lib/SIL/SILProfiler.cpp]: Thanks for pointing this out, some nice simplifications fall out of not having special handling for autoclosures.

File: lib/SIL/SILCoverageMap.cpp (hunk 3 above)
Comment (vedantk) [reply lib/SIL/SILCoverageMap.cpp]: I'm going to look into the fallout from the source compatibility suite now. Let me know if you think I should make the DenseMap change in this PR (I suppose it'd have to be MapVector for deterministic printing).

**Comment by vedantk** [comment]
Ok, all of the source compatibility suite failures should be addressed.

**Comment by vedantk** [comment]
@swift-ci Please smoke test

**Comment by vedantk** [comment]
@swift-ci Please smoke test

**Comment by vedantk** [comment]
@swift-ci Please Test Source Compatibility

**Comment by vedantk** [comment]
@swift-ci Please Test Source Compatibility

**Comment by vedantk** [comment]
@swift-ci Please clean smoke test

**Pull request merged**